contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,248 @@
+
+    /// @dev Stakes the service.
+    /// @param serviceId Service Id.
+    function stake(uint256 serviceId) external {
+        // Check the service conditions for staking
+        (, address multisig, , , , , uint8 state) = IService(serviceRegistry).mapServices(serviceId);
+        // The service must be deployed
+        if (state != 4) {
+            revert WrongServiceState(state, serviceId);
+        }
+        // Check the service security deposit and token, if applicable
+        _checkTokenSecurityDeposit(serviceId);
+
+        // Transfer the service for staking
+        IService(serviceRegistry).transferFrom(msg.sender, address(this), serviceId);
+
+        // ServiceInfo struct will be an empty one since otherwise the transferFrom above would fail
+        ServiceInfo storage sInfo = mapServiceInfo[serviceId];
+        sInfo.multisig = multisig;
+        sInfo.owner = msg.sender;
+        sInfo.nonce = IMultisig(multisig).nonce();
+        sInfo.tsStart = block.timestamp;
+
+        // Add the service Id to the set of staked services
+        setServiceIds.push(serviceId);
+
+        emit ServiceStaked(serviceId, msg.sender);
+    }
+
+    /// @dev Checkpoint to allocate rewards up until a current time.
+    /// @param serviceId Service Id that unstakes, or 0 if the function is called during the deposit of new funds.
+    function _checkpoint(uint256 serviceId) internal returns (uint256 idx) {
+        // Get the service Id set length
+        uint256 size = setServiceIds.length;
+        uint256 lastAvailableRewards = availableRewards;
+        uint256 tsCheckpointLast = tsCheckpoint;
+
+        // If available rewards are not zero, proceed with staking calculation
+        // Otherwise, just bump the timestamp of last checkpoint
+        if (lastAvailableRewards > 0) {
+            uint256 numServices;
+            uint256[] memory eligibleServiceIds = new uint256[](size);
+
+            // Calculate each staked service reward eligibility
+            for (uint256 i = 0; i < size; ++i) {
+                // Get the current service Id
+                uint256 curServiceId = setServiceIds[i];
+
+                // Get the service info
+                ServiceInfo storage curInfo = mapServiceInfo[curServiceId];
+
+                // Calculate the staking nonce ratio
+                uint256 curNonce = IMultisig(curInfo.multisig).nonce();
+                // Get the last service checkpoint: staking start time or the global checkpoint timestamp
+                uint256 serviceCheckpoint = tsCheckpointLast;
+                // Adjust the service checkpoint time if the service was staking less than the current staking period
+                if (curInfo.tsStart > tsCheckpointLast) {
+                    serviceCheckpoint = curInfo.tsStart;
+                }
+                // Calculate the nonce ratio in 1e18 value
+                uint256 ratio = ((block.timestamp - serviceCheckpoint) * 1e18) / (curNonce - curInfo.nonce);
```

Comment:
this is the wrong way round. More tx is more liveness!

Reply:
Also prevent division by zero

Reply:
I can see an issue here with closeness of checkpoints leading to rounding issues

Reply:
This can be defined in either direction. If we believe there will be more tx-s than seconds, then this must be reversed. We are still checking on a specific period of time (from last checkpoint till now) and how many tx-s were performed by every staking service within that period.

Reply:
Agree, incorrect.
It should be tps with dimension: nonce/s (tx/s). In the code we got the inverse value.